### PR TITLE
делает первый ПР

### PR DESCRIPTION
**Pull request by LiliiaShemet**
Проверьте, пожалуйста

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -0,0 +1 @@
+node_modules
```
Comment (Prykhodko-Pasha) [inline .eslintignore:1]: What is this file for?

File: .eslintrc.js [hunk 2]
```diff
@@ -1,12 +1,23 @@
 module.exports = {
+  parser: '@babel/eslint-parser',
```
Comment (Prykhodko-Pasha) [inline .eslintrc.js:2]: What is this line for?

File: .eslintrc.js [hunk 3]
```diff
@@ -1,12 +1,23 @@
...
   env: {
     commonjs: true,
     es2021: true,
     node: true,
+    // 'jest/globals': true,
   },
-  extends: ["standard", "prettier"],
+  extends: [
+    'standard',
+    'plugin:json/recommended',
```
Comment (Prykhodko-Pasha) [inline .eslintrc.js:11]: What is this line for?

File: .eslintrc.js [hunk 4]
```diff
@@ -1,12 +1,23 @@
...
+    'plugin:jest/recommended',
```
Comment (Prykhodko-Pasha) [inline .eslintrc.js:12]: What is this line for?

File: .eslintrc.js [hunk 5]
```diff
@@ -1,12 +1,23 @@
...
+    'prettier',
+  ],
   parserOptions: {
     ecmaVersion: 12,
   },
-  rules: { quotes: ["error", "double"] },
-};
+  rules: {
+    'jest/no-mocks-import': 'off',
```
Comment (Prykhodko-Pasha) [inline .eslintrc.js:19]: What is this line for?

File: .eslintrc.js [hunk 6]
```diff
@@ -1,12 +1,23 @@
...
+  extends: [
+    'standard',
+    'plugin:json/recommended',
+    'plugin:jest/recommended',
+    'prettier',
+  ],
   parserOptions: {
     ecmaVersion: 12,
   },
-  rules: { quotes: ["error", "double"] },
```
Comment (Prykhodko-Pasha) [inline .eslintrc.js:11]: Why did you delete this line?

File: README.md [hunk 7]
```diff
@@ -1 +1,67 @@
-# wallet-backend
+# Final-project
```
Comment (Prykhodko-Pasha) [inline README.md:1]: This description doesn't corresponding to our project

File: .prettierrc [hunk 8]
```diff
@@ -0,0 +1,6 @@
+{
+"trailingComma": "es5",
+"tabWidth": 2,
+"semi": false,
+"singleQuote": true
+}
```
Comment (Prykhodko-Pasha) [inline .prettierrc:6]: Some rules are incorrect for our project. Please delete this file or change these rules to: "trailingComma": "all", "semi": true

File: app.js [hunk 9]
```diff
@@ -1,32 +1,76 @@
-const express = require("express");
-const logger = require("morgan");
-const cors = require("cors");
-require("dotenv").config();
-
+const express = require('express')
+const cookieParser = require('cookie-parser')
+// const path = require('path')
+const swaggerUi = require('swagger-ui-express')
```
Comment (Prykhodko-Pasha) [inline app.js:4]: What is this package for?

File: app.js [hunk 10]
```diff
@@ -1,32 +1,76 @@
...
+const swaggerDocument = require('./swagger.json')
```
Comment (Prykhodko-Pasha) [inline app.js:5]: What is this line for? I can't see this file in our project

File: app.js [hunk 11]
```diff
@@ -1,32 +1,76 @@
...
+const helmet = require('helmet')
```
Comment (Prykhodko-Pasha) [inline app.js:6]: What is this package for?

File: app.js [hunk 12]
```diff
@@ -1,32 +1,76 @@
...
+const cors = require('cors')
+const logger = require('morgan')
+const rateLimit = require('express-rate-limit')
+// const swaggerUi = require('swagger-ui-express')
+// const swaggerDocument = require('../../swagger.json')
+const { apiLimit, jsonLimit } = require('./config/rate-limit.json')
+const { HttpCode } = require('./helpers/constants')
+require('dotenv').config()
+const authRouter = require('./routes/api/auth')
```
Comment (Prykhodko-Pasha) [inline app.js:15]: We don't have such route in our project. Please follow the swagger file https://app.swaggerhub.com/apis-docs/Prykhodko-Pasha/wallet/1.0.0#/

File: app.js [hunk 13]
```diff
@@ -1,32 +1,76 @@
...
 
+const formatsLogger = app.get('env') === 'development' ? 'dev' : 'short'
+app.get('env') !== 'test' && app.use(logger(formatsLogger))
+app.use(
+  cors({
+    credentials: true,
+    origin:
+      process.env.NODE_ENV === 'development'
+        ? process.env.BASE_URL
+        : process.env.CLIENT_URL,
```
Comment (Prykhodko-Pasha) [inline app.js:33]: We don't have such variables in our env file

File: app.js [hunk 14]
```diff
@@ -1,32 +1,76 @@
...
 app.use(logger(formatsLogger));
-app.use(cors());
-app.use(express.json());
 // app.use(express.static("public"));
-
 // app.use("/api/users", usersRouter);
 app.use("/api/transactions", transactionsRouter);
 app.use("/api/categories", categoriesRouter);
+app.use(cookieParser())
+app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument))
```
Comment (Prykhodko-Pasha) [inline app.js:46]: We don't have such route in our project

File: app.js [hunk 15]
```diff
@@ -1,32 +1,76 @@
...
+app.use('/auth', authRouter)
```
Comment (Prykhodko-Pasha) [inline app.js:47]: We don't have such route in our project

File: app.js [hunk 16]
```diff
@@ -1,32 +1,76 @@
...
+app.use('/users', usersRouter)
+app.use('/test', testingRouter)
```
Comment (Prykhodko-Pasha) [inline app.js:49]: We don't have such route in our project

File: app.js [hunk 17]
```diff
@@ -1,32 +1,76 @@
...
+  res.status(404).json({ message: 'Not found' })
+})
 
-app.use((err, req, res, next) => {
-  const { status = 500, message = "Server error" } = err;
-  res.status(status).json({ message });
-});
+app.use((err, _req, res, _next) => {
+  res.status(err.status || 500).json({ message: err.message })
+})
```
Comment (Prykhodko-Pasha) [inline app.js:74]: Why did you change this app.use?

File: babel.config.json [hunk 18]
```diff
@@ -0,0 +1,7 @@
+{
+  "plugins": [
+    "@babel/plugin-syntax-top-level-await",
+    "@babel/plugin-proposal-class-properties",
+    "@babel/plugin-proposal-private-methods"
+  ]
+}
```
Comment (Prykhodko-Pasha) [inline babel.config.json:7]: What is this file for?

File: model/db.js [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+  console.log('Mongoose disconnected')
+})
+
+process.on('SIGINT', async () => {
+  await mongoose.connection.close()
+  console.log('Connection for db closed and terminate')
+  process.exit(1)
+})
+
+module.exports = db
```
Comment (Prykhodko-Pasha) [inline model/db.js:30]: Could you explain what is this file for?

File: config/email.json [hunk 20]
```diff
@@ -0,0 +1,4 @@
+{
+  "dev": "http://localhost:3000",
```
Comment (Prykhodko-Pasha) [inline config/email.json:2]: We need to launch our backend at the 3001 port, because our frontend will be launched at 3000 port

File: config/email.json [hunk 21]
```diff
@@ -0,0 +1,4 @@
...
+  "prod": "https://wallet-backend.herokuapp.com"
```
Comment (Prykhodko-Pasha) [inline config/email.json:3]: You already have deployed our backend on Heroku?

File: controllers/testing.js [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+  } catch (e) {
+    next(e)
+  }
+}
+
+module.exports = {
+  getTechQuestions,
+  getTheoryQuestions,
+  postResult,
+}
```
Comment (Prykhodko-Pasha) [inline controllers/testing.js:80]: What is this file for?

File: controllers/users.js [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+  }
+  return result
+}
+
+module.exports = {
+  updateUserName,
+  getCurrent,
+  avatars,
+  verify,
+}
```
Comment (Prykhodko-Pasha) [inline controllers/users.js:161]: See what controllers have already exist - you do double work

File: helpers/create-dir.js [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+    .then(() => true)
+    .catch(() => false)
+}
+const createFolderIsNotExist = async (folder) => {
+  if (!(await isAccessible(folder))) {
+    await fs.mkdir(folder, { recursive: true })
+  }
+}
+
+module.exports = createFolderIsNotExist
```
Comment (Prykhodko-Pasha) [inline helpers/create-dir.js:15]: What is this file for?

File: helpers/errorHandler.js [hunk 25]
```diff
@@ -0,0 +1,12 @@
...
+    super()
+    this.status = status
+    this.message = message
+    this.data = data
+  }
+}
+
+module.exports = {
+  ErrorHandler,
+}
```
Comment (Prykhodko-Pasha) [inline helpers/errorHandler.js:12]: I can't see where do you use this class

File: helpers/upload.js [hunk 26]
```diff
@@ -0,0 +1,27 @@
+const multer = require('multer')
+const path = require('path')
+require('dotenv').config()
+const UPLOAD_DIR = path.join(process.cwd(), process.env.UPLOAD_DIR_NAME)
```
Comment (Prykhodko-Pasha) [inline helpers/upload.js:4]: We don't have UPLOAD_DIR_NAME in our env file

File: jest.config.js [hunk 27]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  testEnvironment: 'node',
+}
```
Comment (Prykhodko-Pasha) [inline jest.config.js:3]: What is this file for?

File: model/db.json [hunk 28]
```diff
@@ -0,0 +1,776 @@
...
+        "Component testing",
+        "Acceptance testing",
+        "Performance testing",
+        "All options are correct",
+        "All options are incorrent"
+      ],
+      "rightAnswer": "Usability testing"
+    }
+  ]
+}
```
Comment (Prykhodko-Pasha) [inline model/db.json:776]: WHAT THE FUCK IS THIS?

File: model/schemas/techQuestion.js [hunk 29]
```diff
@@ -0,0 +1,18 @@
...
+  },
+  {
+    versionKey: false,
+    timestamps: true,
+  }
+)
+
+const TechQuestion = model('tech-question', techQuestionSchema)
+
+module.exports = TechQuestion
```
Comment (Prykhodko-Pasha) [inline model/schemas/techQuestion.js:18]: What is this file for?

File: model/schemas/theoryQuestion.js [hunk 30]
```diff
@@ -0,0 +1,18 @@
...
+  },
+  {
+    versionKey: false,
+    timestamps: true,
+  }
+)
+
+const TheoryQuestion = model('theory-question', theoryQuestionSchema)
+
+module.exports = TheoryQuestion
```
Comment (Prykhodko-Pasha) [inline model/schemas/theoryQuestion.js:18]: What is this file for?

File: model/schemas/user.js [hunk 31]
```diff
@@ -0,0 +1,68 @@
+const { Schema, model } = require('mongoose')
+const bcrypt = require('bcryptjs')
+const SALT_FACTOR = 6
+const gravatar = require('gravatar')
+
+const userSchema = new Schema(
+  {
+    name: {
+      type: String,
+      default: null,
```
Comment (Prykhodko-Pasha) [inline model/schemas/user.js:10]: Isn't this field required?

File: model/schemas/user.js [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+// 	return re.test(String(value).toLowerCase())
+// })
+
+userSchema.methods.validPassword = async function (password) {
+  return await bcrypt.compare(password, this.password)
+}
+
+const User = model('user', userSchema)
+
+module.exports = User
```
Comment (Prykhodko-Pasha) [inline model/schemas/user.js:68]: Please move this file to the correct folder

File: model/testing.js [hunk 33]
```diff
@@ -0,0 +1,37 @@
...
+  const questions = await TechQuestion.find()
+  return questions
+}
+
+module.exports = {
+  listTechQuestions,
+  listTheoryQuestions,
+  listAnswersTechQuestions,
+  listAnswersTheoryQuestions,
+}
```
Comment (Prykhodko-Pasha) [inline model/testing.js:37]: What is this file for?

File: model/users.js [hunk 34]
```diff
@@ -0,0 +1,56 @@
...
+
+module.exports = {
+  createSession,
+  findSession,
+  deleteSession,
+  findUserByField,
+  create,
+  updateAvatar,
+  updateUserByField,
+}
```
Comment (Prykhodko-Pasha) [inline model/users.js:56]: This is not a model file

File: package.json [hunk 35]
```diff
@@ -1,35 +1,58 @@
 {
-  "name": "wallet-backend",
-  "version": "1.0.0",
+  "name": "final_project",
+  "version": "0.0.0",
```
Comment (Prykhodko-Pasha) [inline package.json:3]: Why did you change the name and version?

File: package.json [hunk 36]
```diff
@@ -1,35 +1,58 @@
...
+    "eslint-plugin-json": "2.1.2",
     "eslint-plugin-node": "^11.1.0",
-    "eslint-plugin-promise": "5.2.0",
-    "nodemon": "2.0.15"
+    "eslint-plugin-promise": "^4.2.1",
+    "jest": "26.6.3",
+    "nodemon": "^2.0.7",
+    "supertest": "6.1.3"
   }
 }
```
Comment (Prykhodko-Pasha) [inline package.json:58]: Please delete all the carets and tildas

File: routes/api/auth/index.js [hunk 37]
```diff
@@ -0,0 +1,28 @@
...
+
+router.get('/google', tryCatchWrapper(authController.googleAuth))
+router.get('/google-redirect', tryCatchWrapper(authController.googleRedirect))
+router.get(
+  '/refresh-token',
+  tryCatchWrapper(validateRefreshToken),
+  tryCatchWrapper(authController.refreshToken)
+)
+
+module.exports = router
```
Comment (Prykhodko-Pasha) [inline routes/api/auth/index.js:28]: we don't have such routes in our project

File: routes/api/auth/validation.js [hunk 38]
```diff
@@ -0,0 +1,45 @@
...
+  next()
+}
+
+module.exports.Register = (req, res, next) => {
+  return validate(schemaRegister, req.body, next)
+}
+
+module.exports.Login = (req, res, next) => {
+  return validate(schemaLogin, req.body, next)
+}
```
Comment (Prykhodko-Pasha) [inline routes/api/auth/validation.js:45]: This is not a routes file

File: routes/api/testing/index.js [hunk 39]
```diff
@@ -0,0 +1,10 @@
+const express = require('express')
+const router = express.Router()
+const testingController = require('../../../controllers/testing')
+const guard = require('../../../helpers/guard')
+// const validation = require('./validation')
+router.post('/result', guard, testingController.postResult)
+router.get('/tech-questions', guard, testingController.getTechQuestions)
+router.get('/theory-questions', guard, testingController.getTheoryQuestions)
+
+module.exports = router
```
Comment (Prykhodko-Pasha) [inline routes/api/testing/index.js:10]: we don't have such routes in our project

File: routes/api/users/index.js [hunk 40]
```diff
@@ -0,0 +1,25 @@
...
+router.patch(
+  '/current',
+  guard,
+  validation.UpdateUser,
+  usersController.updateUserName
+)
+
+router.get('/verify/:verificationToken', usersController.verify)
+
+module.exports = router
```
Comment (Prykhodko-Pasha) [inline routes/api/users/index.js:25]: we don't have such routes in our project

File: routes/api/users/validation.js [hunk 41]
```diff
@@ -0,0 +1,35 @@
...
+    next(
+      res.status(HttpCode.BAD_REQUEST).json({
+        status: 'error',
+        code: HttpCode.BAD_REQUEST,
+        message: 'Field AvatarURL with file is not found',
+      })
+    )
+  }
+  next()
+}
```
Comment (Prykhodko-Pasha) [inline routes/api/users/validation.js:35]: This is not a routes file

File: services/email.js [hunk 42]
```diff
@@ -0,0 +1,66 @@
...
+      theme: 'cerberus',
+      product: {
+        name: 'Contacts',
+        link: this.link,
+      },
+    })
+    const template = {
+      body: {
+        name,
+        intro: 'Need Verification for QA TEST WEB APP',
```
Comment (Prykhodko-Pasha) [inline services/email.js:37]: QA TEST WEB APP ?

**Comment by LiliiaShemet** [comment]
Дякую за перевірку. Іду правити. Врахую всі твої зауваження,  перероблю маршрути.
